[VideoVersions] Fix/improve the generation/import of versions.

78andyp opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Thanks to @ksooo and @CrystalP for comments on movie versions in previous commits.

**Commit 1**

As suggested by @CrystalP, remove the use of a FileItem property.

**Commit 2**

Thanks to @ksooo for pointing out `existingMovieId` was unused.
This was a result of the second fix of #27129.
The side effect of this was versions spread accross nfos were not properly added if they had a different filename/directory derived title to the one in the nfo. New movie entries were created rather than video version entries.
This commit fixes that.

**Commit 3**

While investigating above I was looking at how Disc n is appended to titles for bluray/dvds.
This makes sense when there are multiple discs containing the same movie.
For example - A lot of UHD packages also contain a bluray disc of the movie (often with additional extras etc..) - and these would often be placed in subdirectories `Disc n` under the main movie directory eg. `Aliens (1986)`
To avoid duplicate entries in the library #24076 appended `(Disc n)` to the title and created a set (if the movie wasn't in one already).

However, when converting to versions the `(Disc n)` is not needed any more so this commit removes it.

**Commit 4**

As suggested by @CrystalP use AddVideoAsset() to add a version.

**Commit 5**

When importing movie versions from nfos the art assigned to the movie is that of the first version added, however it may not be the default version - so when the default version is set at the end, make sure the art is updated as well.

**Commit 6**

`AddNewMovie()` assumed all versions were the default (40400). Now, if an asset (version) id is specified in the VideoInfoTag() that is used.

When user defined video version ids (eg. a specific type of edit) then that video version needs adding to the videoversiontype table (if it doesn't already exist) in AddNewMovie() before the version itself is created.

**Commit 7**

Bluray versions, when created as additional versions on a disc, have no art but probably should have the same art as the base version (at least to start with).

**Commit 8**

Remove definition for a function that doesn't exist

**Commit 9**

Improved logging for finding versions (or not) in a multi-movie nfo.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Issues identified above.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Locally and on slack.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] All new and existing tests passed

78andyp committed:
Fix add movie versions from nfos when title is different to that derived from the file/directory.

78andyp committed:
Remove (Disc n) from movie titles when multiple discs are combined into movie versions. Add tests.

78andyp committed:
Refactor nfo import movie version code to use AddVideoAsset() and AddOrUpdateVideoVersion().

78andyp committed:
Ensure art is set for default movie version when importing.

78andyp committed:
Set video version id in AddNewMovie() rather than assuming default, adding custom ids as needed (both for new movies and new movie versions).

78andyp committed:
Remove property idMovie.

78andyp committed:
Duplicate art for bluray versions.

78andyp committed:
Remove function with no definition.

78andyp committed:
Improve log messaging when scanning for multiple versions in single nfo.

CrystalP commented:
I'm trying to get into this, is the description of the PR still truthful to the content of the commits? There are ten of them now, not 6.
Is it still attempting to fix #27130? Wasn't it taken care of by another merged PR?

CrystalP commented on xbmc/video/VideoInfoScanner.cpp:
```diff
@@ -978,29 +978,43 @@ CVideoInfoScanner::~CVideoInfoScanner()
+        // See if movie already in library
+        // Try with title first, then path
+        const std::string title{tag->GetTitle()};
+        const int uniqueId{std::atoi(tag->GetUniqueID(tag->GetDefaultUniqueID()).c_str())};
```
Nevermind that copilot comment, uniqueid is not of integer type. And GetUniqueID() is a shortcut for GetUniqueID(GetDefaultUniqueID()).
The function feels like duplication of the existing GetSameVideoItems()

78andyp replied:
Code rewritten

78andyp commented:
Helpful comments from @CrystalP taken from slack - I've put here as I keep losing them on there!

```
After reading some of the code I understand why. Using the isdefaultvideoversion tag is ok, it's the only way.
However the nfo import (from library import or library scan of standalone nfo) cannot rely on the hasvideoversions tag. It shouldn't even be exported in the first place.
That's because it's a calculated field that by definition depends on the existence of other versions than the one being processed. It's not "standalone" for a given version.
Relying on it causes two problems I could think of:
- for the most common case of adding a new movie/version to the library:
The user (or a nfo generating media manager) has to know the current contents of the library to set the correct hasversions value in the new nfo accompanying the new video.
Not practical/acceptable as far as I am concerned. And the value in nfo gets out of date if the user makes changes in the library after creating the nfo.
- for a library export / import: as a backup mechanism, they typically don't happen back to back so the user can make changes to the files in the video sources, and add/remove movies/versions, invalidating the exported hasversions values since the nfos won't change at the same time.
Importing the outdated nfo may yield incorrect results (depends on exactly what the code does with hasversions).
Other point, the trick of checking lastTitle == currentTitle in ImportFromXML is highly suspicious and will cause problems.
First, related movie versions cannot be assumed to be processed consecutively (user may have one source with HD, another one on different disc with UHD versions)
Second, even when that's the case, the title is not a proper way to uniquely identify a movie, as I think [@karellen](https://teamkodi.slack.com/team/U5DCTJU6N) found out in testing.
There is an existing function (CVideoDatabase::GetSameVideoItems) to properly check the existence of a movie in the library using the unique ids first then movie and year.
```

78andyp commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -324,6 +326,29 @@ bool CGUIDialogVideoManagerVersions::AddVideoVersion()
+  if (const int offset{GetPartNumberInTitle(title)}; offset >= 0)
+  {
+    title.erase(offset);
+    tag->m_strTitle = title;
```
Done.

78andyp commented on xbmc/utils/URIUtils.cpp:
```diff
@@ -674,6 +674,23 @@ int URIUtils::GetBlurayPlaylistFromPath(const std::string& path)
+{
+  // Build regex inserting local specific spelling of disc (xxx)
+  // [ _.-]*\((?:xxx|dis[ck])[ _.-]*\d{1,3}\)$
+  std::string localeDiscStr{StringUtils::Format("{} ", g_localizeStrings.Get(427))};
+  if (!localeDiscStr.empty())
+    localeDiscStr += "|";
```
Done

78andyp commented:
> I'm trying to get into this, is the description of the PR still truthful to the content of the commits? There are ten of them now, not 6. Is it still attempting to fix #27130? Wasn't it taken care of by another merged PR?

Updated.

78andyp commented:
@CrystalP - I've updated the code to use GetSameVideoItems.

This needed some tweaking as it's predicated on the item in question already existing in the database. In this case that isn't true - it hasn't been added yet - to the uniqueid is extracted from the tag.

Also, at this stage, we don't want matches just based on title - as that could mean Superman (1978) and Superman (2025) are considered the same movie.

It also didn't use path based matching - which can be more specific than the above (ie - if movies are in named folders - assume that any movie found within that folder is a version).

So I've added an options mask. This should mean there is no functionality change for your (the default) use cases but can accomodate this.

I'm working through your other two points - using HasVersions() and the comparison.

Thanks
